calico/calcollections.py

```diff
@@ -75,3 +75,89 @@ def resulting_size(self):
     @property
     def empty(self):
         return not (self.added_entries or self.removed_entries)
+
+
+class OneToManyIndex(object):
+    """
+    Represents a mapping from key to a set of values.
+    """
+
+    def __init__(self, set_cls=set):
+        self._set_cls = set_cls
+        self._index = {}
+
+    def add(self, key, value):
+        """Add a mapping to the index.
+
+        Idempotent: does nothing if the mapping is already present.
+        """
+        index_entry = self._index.setdefault(key, value)
+        if index_entry != value:
+            # Failed to insert the new value as the single entry, examine
+            # what we got.
+            if isinstance(index_entry, self._set_cls):
+                # Already have multiple values for that entry, add the new one.
+                index_entry.add(value)
+            else:
+                # There was an entry but it wasn't the one we tried to add,
+                # promote the entry to a set.
+                index_entry = self._set_cls([index_entry, value])
+                self._index[key] = index_entry
```

Comment:
This is an occupancy optimization (that I benchmarked and tested for the label index back when we were scale testing).  Added a comment to the class.